core/ibft.go

```diff
@@ -0,0 +1,1134 @@
+			})
+		}
+	}
+
+	if len(roundsAndPreparedBlockHashes) == 0 {
+		return true
+	}
+
+	// Find the max round
+	var (
+		maxRound     uint64 = 0
+		expectedHash []byte = nil
+	)
+
+	for _, tuple := range roundsAndPreparedBlockHashes {
+		if tuple.round > maxRound {
+			maxRound = tuple.round
+			expectedHash = tuple.hash
+		}
+	}
+
+	return bytes.Equal(expectedHash, proposalHash)
+}
+
+//	handlePrePrepare parses the received proposal and performs
+//	a transition to PREPARE state, if the proposal is valid
+func (i *IBFT) handlePrePrepare(view *proto.View) *proto.Message {
+	isValidPrePrepare := func(message *proto.Message) bool {
+		if view.Round == 0 {
+			//	proposal must be for round 0
+			return i.validateProposal0(message, view)
+		}
+
+		return i.validateProposal(message, view)
+	}
+
+	msgs := i.messages.GetValidMessages(
+		view,
+		proto.MessageType_PREPREPARE,
+		isValidPrePrepare,
+	)
+
+	if len(msgs) < 1 {
+		return nil
+	}
+
+	return msgs[0]
+}
+
+// runPrepare runs the Prepare IBFT state
+func (i *IBFT) runPrepare(ctx context.Context) error {
+	i.log.Debug("enter: prepare state")
+	defer i.log.Debug("exit: prepare state")
+
+	var (
+		// Grab the current view
+		view = i.state.getView()
+
+		// Grab quorum information
+		quorum = i.backend.Quorum(view.Height)
+
+		// Subscribe to PREPARE messages
+		sub = i.messages.Subscribe(
+			messages.SubscriptionDetails{
+				MessageType: proto.MessageType_PREPARE,
+				View:        view,
+				NumMessages: int(quorum) - 1,
+			},
+		)
+	)
+
+	// The subscription is not needed anymore after
+	// this state is done executing
+	defer i.messages.Unsubscribe(sub.ID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Stop signal received, exit
+			return errTimeoutExpired
+		case <-sub.SubCh:
+			if !i.handlePrepare(view, quorum) {
+				//	quorum of valid prepare messages not received, retry
+				continue
+			}
+
+			return nil
+		}
+	}
+}
+
+//	handlePrepare parses available prepare messages and performs
+//	a transition to COMMIT state, if quorum was reached
+func (i *IBFT) handlePrepare(view *proto.View, quorum uint64) bool {
+	isValidPrepare := func(message *proto.Message) bool {
+		// Verify that the proposal hash is valid
+		return i.backend.IsValidProposalHash(
+			i.state.getProposal(),
+			messages.ExtractPrepareHash(message),
+		)
+	}
+
+	prepareMessages := i.messages.GetValidMessages(
+		view,
+		proto.MessageType_PREPARE,
+		isValidPrepare,
+	)
+
+	if len(prepareMessages) < int(quorum)-1 {
+		//	quorum not reached, keep polling
+		return false
+	}
+
+	// Multicast the COMMIT message
+	i.transport.Multicast(
+		i.backend.BuildCommitMessage(
+			i.state.getProposalHash(),
+			view,
+		),
+	)
+
+	i.log.Debug("commit message multicasted")
+
+	i.state.setLatestPC(&proto.PreparedCertificate{
+		ProposalMessage: i.state.getProposalMessage(),
+		PrepareMessages: prepareMessages,
+	})
+
+	i.state.setLatestPPB(i.state.getProposal())
+
+	// Move to the commit state
+	i.state.changeState(commit)
```

Comment:
Also state should have a method for a specific usecase because it's easier to understand why state changes.